Newcolor

Neon22 opened the pull request:
Used your idea to add Color class as namedtuple.

Left my hack for table bg pattern as is but using new Color()

Neon22 committed:
Adding background cell color

modified rtf.py to allow for add_color which returns index of the rgb for use on a row background

modified table to allow addition of bgtable color pattern.

also cleaned up my additions a little

Neon22 committed:
change to Color class

adjusted so tests pass.
uses new Color named tuple class

Neon22 commented on rtflib/rtf.py:
```diff
@@ -130,7 +128,7 @@ def rtf_code(self) -> str:
             code += "\\colortbl;\n"
             for color in self.colors:
-                code += f"\\red{str(color[0])}\\green{str(color[1])}\\blue{str(color[2])};\n"
+                code += color.rtf_code#code += f"\\red{str(color[0])}\\green{str(color[1])}\\blue{str(color[2])};\n"
```
forgot to remove this comment

Neon22 committed:
oops comment removed

viking-sudo-rm commented on rtflib/rtf.py:
```diff
@@ -23,10 +23,24 @@
 """
 
 from abc import ABCMeta, abstractmethod
+from collections import namedtuple
```
I suggest using `from typing import NamedTuple` since it allows a typed named tuple

viking-sudo-rm commented on rtflib/rtf.py:
```diff
@@ -81,30 +94,29 @@ def __init__(self):
     def add(self, element: RtfElement) -> None:
         if element.iscompatible("rtf"):
             self.elements.append(element)
             if "color" in dir(element) and element.color:
```
It would be cleaner to get rid of the condition in 115 here. Instead, we should expect that the user has already registered the color that is passed as an argument with `rtf.add_color`:
```python
line = Line("hello world", color=rtf.add_color(255, 0, 0))
```
I propose changing the code this way and updating the examples and tests to reflect it

Also, replace `"color" in dir(element)` with `hasattr(element, "color")` (although this might get removed)

Neon22 replied:
ok. so if the color does not already exist then inline the rtf.add_color() and rely on `add_color()` to return the old color if this new one already exists ?

viking-sudo-rm commented on rtflib/tests/test_rtflib.py:
```diff
@@ -75,5 +75,38 @@ def test_table_borders(self):
+        rtf.preelements.append(page_orientation)
+        rtf.add(Line("hello world. A Line of text.\n"))
+        rtf.add(Line("hello world. A Line of text.\n", format=Format(bold=True,strike=True,size=16)))
+        rtf.add(Line("hello world. A Red line\n", color=Color(255, 0, 0)))
```
See other comment. I propose we change the API so that the user should register the color here with `rtf.add_color(..)`. Then each `Color` object will store a color ID internally

Neon22 replied:
do you mean the user would do something like:
```
rtf = Rtf()
myred = rtf.add_color(255, 0, 0)
rtf.add(Line("hello world. A Red line\n", color=myred))
```
so myred would be a Color()

viking-sudo-rm reviewed (commented):
I left some comments about how `rtf.add_color` can be used instead of the `Color` constructor when creating `Line`s.

I also suggest using `typing.NamedTuple` instead of `collections.namedtuple`

viking-sudo-rm commented:
For backward compatibility, we can also leave the current condition that calls add_color if it is also that case that color.color_id is None. Otherwise, it can use the already registered color.

This way, the current API won't break but also it will be possible for users to register colors and pass references to them themselves (which is a cleaner design IMO).

Does this make sense?

Neon22 committed:
updated Color mechanism

uses rtf.add_color() instead of Color()
test generator file updated.

Neon22 commented:
so:
- now using typed.NamedTuple
  - (note that according to PEP it does not check types...)
- changed dir(element) to hasattr as suggested
  - probably do need this as otherwise can add a color to an element that does not support it (as I tried to do when trying to work out how it worked)...
- changed test code to use preferred doc method of rtf.add_color(R,G,B) instead of a simple Color(R,G,B) designation.
  - note that code works either way and resolves uniqueness in both cases.
 
I have not updated the readme.md. Maybe all that is required is to refer users to the tests to see examples of usage ?

viking-sudo-rm commented:
Thanks, this looks great! Two minor comments before merging

1. A pedantic suggestion: could you add two versions of `test_hello_world_red`, one which uses `rtf.add_color(..)` and one which uses `Color(..)` directly? Would be good to explicitly test that both methods work

2. A question: what happens if you pass a color created with `Color(..)` as a table bg color rather than a color created with `rtf.add_color`? The desired behavior should be that it applies the same logic as `Line` (and registers the new color) in my opinion

Neon22 commented:
I can confirm `Color()` and `rtf.add_color()` behave the same in Line().
Alas there is no rtf reference in Table and so the current mechanism of using `rtf.add_color()` is the only mech that works for Tables.

Only workaround I can think of is to add colors to `rtfElement()` and hen gather them all up in `Rtf()` but its getting ugly.

What do you suggest ? (I'm for leaving Table as-is)

viking-sudo-rm commented:
Got it. I think it's fine for now, but just wanted to clarify how it works.

Will merge now, thanks for the contribution!

viking-sudo-rm commented:
Closes #7
